src/lib/prep.ml

```diff
@@ -189,50 +187,45 @@ module Prep = struct
           (fun acc str -> acc || Astring.String.is_infix ~affix:str log_line) false retry_on
       in
       match Storage.parse_hash ~prefix:"HASHES" line with
-      | Some value -> (value :: git_hashes, failed, retriable_errors)
+      | Some value -> ((value :: git_hashes, failed), retriable_errors)
       | None -> (
+        if retry_conditions line then
+          ((git_hashes, failed), line :: retriable_errors)
+        else
           match String.split_on_char ':' line with
           | [ prev; branch ] when Astring.String.is_suffix ~affix:"FAILED" prev ->
               Current.Job.log job "Failed: %s" branch;
-              (git_hashes, branch :: failed, retriable_errors)
-          | [ prev; branch ] when retry_conditions prev ->
-              (git_hashes, failed, branch :: retriable_errors)
-          | _ -> (git_hashes, failed, retriable_errors))
+              ((git_hashes, branch :: failed), retriable_errors)
+          | _ -> ((git_hashes, failed), retriable_errors))
     in
 
+    (* (unit -> (('a * 'c list), 'e) Lwt_result.t) *)
+    let fn () =
+      let* result = Current_ocluster.Connection.run_job ~job build_job in
+      let result = match result with Error _ -> Error() | Ok _ -> Ok(([], []), []) in
+      let** x = Misc.fold_logs build_job (Misc.with_error_check extract_hashes) result in
+      match x with
+      | Ok x -> Lwt.return_ok x
+      | Error () -> Lwt.return_error (`Msg "")
```

Comment:
Same comment here as for `compile.ml` above. This should be preserving an error string.